testutil/ensurelogchecker.go

```diff
@@ -0,0 +1,268 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package testutil
+
+import (
+	"bufio"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"strings"
+
+	"gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/strutil"
+)
+
+// CheckEnsureLoopLogging checks for trace log statement coverage in
+// StateManager implementers.
+// The provided file (*mgr.go) should contain a StateManager Ensure implementation.
+// If the specified file uses submanagers, their source files must be included.
+// If expectChildEnsureMethods, will check that the Ensure method in the go
+// source code, indicated by the given file name, has at least one child
+// ensure method and the file contains at least one trace log inside each
+// ensure* method called within that file's Ensure() method.
+// If not expectChildEnsureMethods, then the go source code must
+// not contain any child ensure methods.
+func CheckEnsureLoopLogging(filename string, c *check.C, expectChildEnsureMethods bool, submanagerFiles ...string) {
+	logTemplate := `logger.Trace("ensure", "manager", "%s", "func", "%s")`
+	parsedMgrFile, err := newParsedFile(filename)
+	c.Assert(err, check.IsNil)
+	childEnsures := ensureCallList(parsedMgrFile.file, childEnsureFunc)
+	if expectChildEnsureMethods {
+		c.Assert(len(childEnsures), IntGreaterThan, 0)
+	} else {
+		c.Assert(len(childEnsures), IntEqual, 0)
```

Comment:
nitpick, there's a `HasLen` checker we can use here
```suggestion
		c.Assert(childEnsures, check.Not(check.HasLen), 0)
	} else {
		c.Assert(childEnsures, check.HasLen, 0)
```